### PR TITLE
fix(invoices): race conditions for wbehook

### PR DESCRIPTION
**Pull request by jdenquin**
## Context

- we had a race condition that was sending the webhook for a invoice before closing the transaction DB for it.

## Description

- Call the webhook job after the invoice DB transaction

### CODE REVIEW

File: app/services/invoices/create_service.rb [hunk 1]
```diff
@@ -32,12 +32,13 @@ def create
...
         ).call
 
-        SendWebhookJob.perform_later(:invoice, invoice) if should_deliver_webhook?
         Invoices::Payments::CreateService.new(invoice).call
         track_invoice_created(invoice)
 
         result
       end
+
+      SendWebhookJob.perform_later(:invoice, invoice) if should_deliver_webhook?
```
Comment (vincent-pochet) [inline app/services/invoices/create_service.rb:41]: I think both `Invoices::Payments::CreateService.new(invoice).call` and `track_invoice_created(invoice)` are in the same situation

And we should keep returning the `result` :)

Comment (jdenquin) [reply app/services/invoices/create_service.rb:41]: oh yes sorry for the `result`, I'll move the other 2 also

File: app/services/invoices/create_service.rb [hunk 2]
```diff
@@ -12,6 +12,9 @@ def initialize(subscriptions:, timestamp:)
     end
 
     def create
+      invoice = nil
+      result = nil
```
Comment (vincent-pochet) [inline app/services/invoices/create_service.rb:16]: Actually, self.result is already defined so, I suggest to only do:

```ruby
  ActiveRecord::Base.transaction do
        result.invoice = Invoice.create!(
   # ...
```
and add a `delegate :invoice, to: :result` in the private section of the class

**Review by vincent-pochet — APPROVED** [review]
It works, but allow me to be "punctilious" :smile:

File: app/services/invoices/create_service.rb (hunk 2 above)
Comment (jdenquin) [reply app/services/invoices/create_service.rb:16]: that works for me!

File: app/services/invoices/create_service.rb [hunk 3]
```diff
@@ -66,4 +66,8 @@ def track_invoice_created(invoice)
       )
     end
   end
+
+  private
```
Comment (jdenquin) [inline app/services/invoices/create_service.rb:70]: Lint/UselessAccessModifier: Useless `private` access modifier.

**Pull request merged**